### ISSUE TITLE
Road to official Twig support

### ISSUE THREAD

**Issue by Toflar** [contao/contao#2638]
Current WIP: https://github.com/contao/contao/pull/2988

Twig integration keeps popping up more and more, not only on GitHub but also during our regular Mumble meetups and with people like @m-vo, we also have devs working on PRs for the core more and more.

It's high time we talked about some general requirements so we can agree on at least some basic goals we want to achieve.
Otherwise we keep having PRs and extensions floating around doing all kinds of stuff, eventually causing lots of energy wasted and frustration generated.

This issue should serve as some general one to refer to and we shall collect all the requirements here. It's designed to be updated regularly.

Requirements:

- [x] Developers coming from the Symfony world should feel comfortable with whatever we're going to integrate. We shouldn't reinvent the wheel. For those using the core-bundle only and not the ME, it shouldn't require any special configuration or feel awkward to them.
- [ ] Not only should it be possible to have the typical `$this->render()` functionality in general Symfony controllers but also our fragment controllers.
- [x] We should be able to override any Contao template with Twig templates. In other words, I should be able to e.g. override `form_row.html5` with `form_row.html.twig`.
- [x] Bundles must be able to override templates of other bundles. The bundle order should matter, like it does for our current template system.
- [x] For Contao 4.*, we need to ensure we have an escaping strategy that already forces people to add the `|raw` filter if they want to output HTML unencoded even if the original value was already encoded (prevent double encoding). This is to teach people how to use Twig properly. Once we come to the point of Contao 5 (or whatever version) and finally get rid of input encoding, that's when this will come in handy.
- [x] Users must be able to override a template in the back end as they are used to now. That doesn't mean we have to necessarily get rid of Twig namespaces or subpaths but the UI should assist in finding the right template and overriding it.
- [x] Overriding a template must not force the user to rebuild the container cache. We need a separate one if that's what the Symfony Twig integration requires by default.

**Comment by richardhj** [comment]
Twig in Shopware supports multi-inheritance:

> **Template multi inheritance**
Due to the plugin and theme system in Shopware 6 it is possible that one storefront template gets extended by multiple plugins or themes, but Twig does not allow multi inheritance out of the box. Therefore we created our own twig functions `sw_extends` and `sw_include`, that work like twigs native extends or include, except that they allow multi inheritance. So it is really important to use the `sw_extends` and `sw_include`, instead of the native `extends` and `include`.

https://docs.shopware.com/en/shopware-platform-dev-en/theme-guide/twig-templates?category=shopware-platform-dev-en/theme-guide

Maybe we can adapt this concept. 🚀 
On the downside, we are promoting non-standard language which is mainly incompatible with IDEs.

**Comment by m-vo** [comment]
I'd like to collect use cases / examples for inheriting and overwriting templates.

Real world examples from current Contao projects as well as things that are currently not possible but would make your life easier are both valuable information! What are you already doing? What would be nice? :slightly_smiling_face: **Help is much appreciated.**

I currently got a vast amount of questions that I like to get sorted. Things like: _What about…_

_…a bundle overwriting something from the core / another bundle / another bundle that overwrites something from another bundle that overwrites something from the core? What should the application be able to control in this chain? …overwriting vs inheriting? What happens if something gets included during this? Are dynamic includes a problem? How are multiple themes in parallel handled? Template variants?_

**Comment by richardhj** [comment]
###  Use case 1:

Support for `{% extends 'fe_page.html.twig' %}`

1. The Contao Core comes with a `fe_page` template. This is the first template in the chain.
2. _Bundle A_ overwrites the `fe_page` because the template is present in `src/Resources/views`. _Bundle B_ can overwrite the `fe_page` from _Bundle A_  the same way. The loading order of the bundles is the determinant, and the Contao Core bundle is usually the first bundle in the loading order.
3. Themes can override the `fe_page` when the template is present in `themes/my_theme/templates` directory (see #2781)
Those templates only are in the template inheritance chain, when the theme is currently active in the frontend.
4. The last possible template in the chain is under `TL_ROOT/templates`.

(I am unsure about the order of 3 and 4.)

> What happens if something gets included during this?

Don't know, let's find out? 😄

**Comment by m-vo** [comment]
### Use case 2:

In Sf you're often just extending a base template. This for instance allows to add head tags by defining the block like so:
```twig
{% block head %}
  {{ parent() }}
  <meta name="foo" content="bar">
{% endblock %}
```

It would be nice to find a solution for this in fragment land where multiple fragments could alter a page controller's base template this way. No idea how, but this is brainstorming time. :brain: :cloud_with_lightning:

**Comment by stefansl** [comment]
Shopware 6 is based on Symfony as well. Their template engine is completely based on blocks, which is nice. 
Pro:
You are able to override blocks without overriding the base template itself
https://docs.shopware.com/de/shopware-6-de/tutorials-und-faq/aenderungen-am-template-vornehmen

Contra:
The templates are quite messed up and finding the right block is often very annoying and time consuming
f.e. https://github.com/shopware/storefront/blob/trunk/Resources/views/storefront/layout/footer/footer.html.twig

**Comment by richardhj** [comment]
> Contra:
> The templates are quite messed up and finding the right block is often very annoying and time consuming
> f.e. https://github.com/shopware/storefront/blob/trunk/Resources/views/storefront/layout/footer/footer.html.twig

I don't think this is "too many blocks" but what I find interesting is the naming of the blocks. In fact, when using `embed` and `extends` in the same template, it might occur that the block names might conflict with each other.

There are two ways to work around this issue:
1.  Use explicit block names (as Shopware does, e.g. `layout_footer_navigation_hotline`),
2. encourage to use aliases (`{% use "blocks.html" with sidebar as base_sidebar, title as base_title %}`)

**Comment by aschempp** [comment]
> It would be nice to find a solution for this in fragment land where multiple fragments could alter a page controller's base template this way. 

Why would a fragment extend a base tag? Also, please remember fragments are separate HTTP request, possibly ESI tags or HInclude (which basically is Turbo Frames).

**Comment by m-vo** [comment]
> Why would a fragment extend a base tag? 

It wouldn't. But there are things in the surrounding context that you may want to control (and would do so typically from inside your templates with overwriting blocks): e.g. adding a script tag or og: metadata or a note in the footer. These things are currently output by the global template the page controller 'owns'. It *feels* like multi-inheritance but could also just be syntactical sugar for a registry. :upside_down_face: 

>  Also, please remember fragments are separate HTTP request, possibly ESI tags or HInclude (which basically is Turbo Frames).

Would still work with static content because templates are compiled before the request is processed. But, yeah, just spitballing. :smile_cat:

**Comment by m-vo** [comment]
I'm going to document my findings so far (WIP)

---
## Tweak twig
### Overwriting and extending
#### Background
In Sf templates are registered under a namespace - typically each bundle has one (`FooBundle`)<sup>*)</sup>  and the app as well (`__main__`). The `FilesystemLoader` then associates file paths with namespace names. There can be multiple paths per namespace, in fact when you're adding a `bundles/FooBundle` directory in your app this path will be prepended to the `FooBundle` namespace. When rendering a template `@FooBundle/bar/baz.html.twig`, the loader will go through all paths in the `FooBundle` namespace and return the first content that's present. So if you added a `baz.html.twig` in your app's `bundles/FooBundle/bar` directory this will be used, otherwise the original one.

<sup>*) actually there is also another `!` prefixed one that is used when a template needs to both inherit and replace a parent - but I'm leaving this out here for brevity</sup>

In **Sf** you typically either create a template for your app or adjust one of a bundle. While bundle `B` *can* extend a template of bundle `A` it will not replace it but provide it under a new namespace/name. (It's possible to prepend `B`'s template path to the template namespace of `A` but then the app has problems overwriting it.) 

In **Contao**, however, template inheritance just follows the extension loading order (+ app last) with no namespaces and an arbitrary length of the 'chain' as long as the name matches. 

#### Scenario
Consider the following scenario where there are bundle `A` (original), bundle `B` (extending A), bundle `C` (also extending A) and the app.
```twig
{# original magic.html.twig template in bundle A  #}
{% block content %}Hello{% endblock %}
```

```twig
{# magic.html.twig template in bundle B  #}
{% extends '@A/magic.html.twig' %}
{% block content %}
  {{ parent() }} world!
{% endblock %}
```

```twig
{# magic.html.twig template in bundle C  #}
{% extends '@A/magic.html.twig' %}
{% block content %}
  <b>{{ parent() }}</b>
{% endblock %}
```

```twig
{# app's magic.html.twig template in bundles/ABundle  #}
{% extends '@A/magic.html.twig' %}
{% block content %}
  <div class="ce_magic">{{ parent() }}</div>
{% endblock %}
```

With a loading order of `A`, `B`, `C` you would expect an output like `<div class="ce_magic"><b>Hello world!</b></div>`.

#### Proposed solution
The challenge is to allow Contao's inheritance model without breaking 'Sf compliance':

- Bundles template namespaces are left untouched 
- Extending templates from within a bundle is possible with the usual `{% extends '@A/magic.html.twig' %}`
- Overwriting templates from within a bundle is indicated by configuration or convention (e.g. templates in a `contao` or `bundles` folder)
- We provide our own `ExtendsTokenParser` and `Loader` and setup the inheritance chain when compiling templates as if the following was put in the templates:

    ```twig
    {# magic.html.twig template in bundle C  #}
    {% extends '@B/magic.html.twig' %} {# <--- note the change! #}
    {% block content %}
      <b>{{ parent() }}</b>
    {% endblock %}
    ```

    ```twig
    {# app's magic.html.twig template in bundles/ABundle  #}
    {% extends '@C/magic.html.twig' %} {# <--- note the change! #}
    {% block content %}
      <div class="ce_magic">{{ parent() }}</div>
    {% endblock %}
    ```

- We also setup a more sophisticated fallback logic when retrieving templates so that the inheritance chain always finds the most appropriate overwritten template (the one under `@C/magic.html.twig` if there was none in the app for instance). 

### Interop

To be able to extend Contao templates from twig templates we could build an interop layer. It's not that straightforward because Contao templates are rendered on the fly while twig templates are compiled. 

I've got a working prototype that roughly does the following:

- Register Contao's html5 templates as twig templates as well
- Add a `NodeVisitor` to the compilation step that alters the block node definitions of a `html5` template and replaces it's body with a special `DisplayProxy` node
- This node will - when compiled - output a call to our interop code to render the template
- Effectively this will pass the call to Contao's existing template logic to render the template but the template's blocks get prepopulated with the content from the twig template.

It's vodoo but it would basically allow doing this (note that we're extending not only overwriting!):

```twig
{# mod_article.html.twig #}
{% extends 'mod_article.html5' %}
{% block content %}
  Look, I wrapped the {{ parent() }} content.
{% endblock %}
```

### Registry and template cache
Both inheritance and interop need a registry and template cache to work. We can control this with our own template loader and we can also work around the issue that - in Sf land - namespace paths are found and registered at container build time (which means a lot of cache clearing trouble when changing things around). 

However, it's a really tough one to be fully compatible with Sf. :upside_down_face:  We're basically shipping an app level functionality as a bundle. The solution to this is probably building everything as opt-in in the core bundle and configuring a chain loader and automagic configuration in the manager-bundle. Phew...

On a side note: We will probably also need to decorate the `Environment` in order to setup everything.

----

coming up:

## Output encoding
## Integration / First class citizen

**Comment by aschempp** [comment]
has this been solved with our experimental Twig support? 🙃

**Comment by m-vo** [comment]
Yes, IMHO we could close this issue and track the further progress in the respective Github project and/or a new issue. 

Things like back end integration, handling of templates + cache defaults in our fragment controllers and so on…
